allauth_2fa/adapter.py

```diff
@@ -46,12 +46,16 @@ def get_2fa_authenticate_url(self, request: HttpRequest) -> str:
 
         # Add "next" parameter to the URL if possible.
         # If the view function smells like a class-based view, we can interrogate it.
-        if getattr(request.resolver_match.func, "view_class", None):
+        if getattr(request, "resolver_match", None) and getattr(
+            request.resolver_match.func,
+            "view_class",
+            None,
+        ):
             view = request.resolver_match.func.view_class()
             view.request = request
             success_url = view.get_success_url()
             query_params = request.GET.copy()
-            if success_url:
+            if success_url and hasattr(view, "redirect_field_name"):
                 query_params[view.redirect_field_name] = success_url
             if query_params:
                 redirect_url += f"?{urlencode(query_params)}"
```

Comment:
I think this is getting complex enough to turn it into a separate free utility function, and in that case we can just use `try: except:`s, something like

```python
def get_next_query_string(request: HttpRequest) -> str | None:
    """
    Get the query string (including the prefix `?`) to redirect to after a successful POST.
    
    If a query string can't be determined, returns None.
    """
    # If the view function smells like a class-based view, we can interrogate it.
    try:
        view = request.resolver_match.func.view_class()
        redirect_field_name = view.redirect_field_name
    except AttributeError:
        # Interrogation failed :(
        return None

    view.request = request
    query_params = request.GET.copy()
    success_url = view.get_success_url()
    if success_url:
        query_params[redirect_field_name] = success_url
    if query_params:
        return f"?{urlencode(query_params)}"
    return None
```

What do you think?

Reply:
Sounds fine by me :) Then we can test that directly.

Reply:
@pbadeer Are you up for implementing the function as @akx suggests?

Reply:
@valberg We had to remove this library from our project so I haven't been keeping eyes on this, sorry. I threw @akx's util into the mix, hopefully this is what you were envisioning.